### ISSUE TITLE
Misleading error message with unnamed lifetimes

### ISSUE THREAD

**Issue by Amanieu** [rust-lang/rust#30790]
This code results in a horrible error message:

``` rust
use std::ops::Deref;
trait Trait {}

struct Struct;

impl Deref for Struct {
    type Target = Trait;
    fn deref(&self) -> &Trait {
        unimplemented!();
    }
}
```

```
<anon>:8:5: 10:6 error: method `deref` has an incompatible type for trait:
 expected bound lifetime parameter ,
    found concrete lifetime [E0053]
<anon>: 8     fn deref(&self) -> &Trait {
<anon>: 9         unimplemented!();
<anon>:10     }
<anon>:8:5: 10:6 help: see the detailed explanation for E0053
```

In particular, note the missing lifetime name on the second line of the error message.

**Comment by GuillaumeGomez** [comment]
If there is no lifetime specified, should we print:

```Shell
A lifetime parameter is needed to bind `&Trait`
```

?

**Comment by Amanieu** [comment]
The correct code in this case is to specify the return value of the function as `&(Trait + 'static)`. What happens here is that `Target = Trait` gets expanded by the compiler into `Target = Trait + 'static`, but the error message doesn't make this clear.

**Comment by estebank** [comment]
Proposed output in #47791:

```
error[E0495]: cannot infer an appropriate lifetime for lifetime parameter in generic type due to conflicting requirements
  --> file.rs:8:5
   |
8  | /     fn deref(&self) -> &Trait {
9  | |         unimplemented!();
10 | |     }
   | |_____^
   |
note: first, the lifetime cannot outlive the anonymous lifetime #1 defined on the method body at 8:5...
  --> file.rs:8:5
   |
8  | /     fn deref(&self) -> &Trait {
9  | |         unimplemented!();
10 | |     }
   | |_____^
   = note: ...so that the method type is compatible with trait:
           expected fn(&Struct) -> &Trait + 'static
              found fn(&Struct) -> &Trait
   = note: but, the lifetime must be valid for the static lifetime...
   = note: ...so that the method type is compatible with trait:
           expected fn(&Struct) -> &Trait + 'static
              found fn(&Struct) -> &Trait
```

Still a bit verbose, but in my opinion a clear readability improvement.
